Mini features completed

jedouard98 opened the pull request:
- Sever version number is now supported with the "server" command
- Uptime is now supported with the "uptime" command
- Tokenized input is now supported in place of the Scanner class

jedouard98 committed:
"Version" request supported

Server version request is now handled in the chat via the command
“version”. Created a method in Context that retrieves a Version #. The
interface BasicView was amended to have a getVersion() method. Added an
instantiated version # in Model.java to represent the version number
along with a method there to access it. The NetworkCode.java file was
amended to have two new codes: get server info request and get server
info response, for the purposes of handling version requests.

jedouard98 committed:
Added "Version" help option

Contributed an addition line to describe help option in the available
help menu

jedouard98 committed:
"Uptime" command supported

Server uptime request is now handled in the chat via the command
“Uptime”. Created a method in Context that retrieves server uptime.

The interface BasicView was amended to have a getUptime() method. Added
an instantiated Time variable in Model.java to represent the time the
server has been up along with a method there to access it.

 The NetworkCode.java file was amended to have two new codes: get
server uptime request and get server uptime response, for the purposes
of handling uptime requests.

jedouard98 committed:
Added Tokenized Input

Added a class to replace the Scanner class that now handles lines of
input from the command line and tokenizes it if input is contained
within quotation marks. It also throws exceptions at mishandled input
and lets the client know in advance.

The Tokenizer class parses through input and inserts it into a list for
later use by Chat.java. It throws exceptions if the input appears to be
unclear, which Chat.java handles with a message.

jedouard98 committed:
Tokenizer update

Reverted Tokenizer to disallowing quotation marks in input not enclosed
in quotation marks

kac320 committed:
Update Tokenizer.java

jedouard98 committed:
Tokenizer Readability Fix

Added additional functions to Tokenizer to help with its readability.
Additionally added more comments for it.

jedouard98 committed:
Uptime print update

Edited it so that when uptime is called in Chat, it returns how long
the server has been running in days hours minutes and seconds. This
included adding a method in the Time class to format seconds to this
format.

Unrelated, deleted the Scanner class imports since they are not being
used anymore!

jedouard98 committed:
Merge remote-tracking branch 'origin/jedouard-mini-features' into jedouard-mini-features

# Conflicts:
#	src/codeu/chat/util/Tokenizer.java

Minor conflict because of dual editing

turtlefarmer commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -46,36 +50,55 @@ public Chat(Context context) {
   // the system wants to exit, the function will return false.
   //
   public boolean handleCommand(String line) {
+    try {
```
Localize the scope of try/catch blocks where possible.

turtlefarmer commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -99,14 +122,18 @@ private Panel createRootPanel(final Context context) {
         System.out.println("  u-sign-in <name>");
         System.out.println("    Sign in as the user with the given name.");
+        System.out.println("  uptime");
+        System.out.println("    Display how long server has been running");
```
Trailing period for consistency

turtlefarmer commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -136,8 +163,8 @@ public void invoke(Scanner args) {
-      public void invoke(Scanner args) {
-        final String name = args.hasNext() ? args.nextLine().trim() : "";
+      public void invoke(Tokenizer args) {
+        final String name = args.hasNext() ? args.next().trim() : "";
```
Now that the code is changed to Tokenizer.next() from Scanner.nextLine(), what should be the behavior if the command:

u-add heng sun

is executed?  It makes sense to me for this to fail since the number of arguments is more than expected.

turtlefarmer commented on src/codeu/chat/client/commandline/Chat.java:
```diff
@@ -207,6 +234,10 @@ public void invoke(Scanner args) {
         System.out.println("    Join the conversation as the current user.");
         System.out.println("  info");
         System.out.println("    Display all info for the current user");
+        System.out.println("  uptime");
```
Consider leaving a TODO comment to fix the duplicate code/messages in the help files in different panels.

turtlefarmer commented on src/codeu/chat/client/core/Context.java:
```diff
@@ -47,4 +48,16 @@ public UserContext create(String name) {
     return users;
   }
+
+  public long Uptime() {
```
camelCase for function names.

turtlefarmer commented on src/codeu/chat/common/NetworkCode.java:
```diff
@@ -18,20 +18,24 @@
-      NEW_CONVERSATION_REQUEST = 13,
-      NEW_CONVERSATION_RESPONSE = 14,
+      GET_SERVER_INFO_REQUEST = 1,
+      GET_SERVER_INFO_RESPONSE = 2,
```
Insert new network types at the end.  This is in general good practice for enum-like data types to make sure they are backwards compatible if they are serialized and stored somewhere.

turtlefarmer commented on src/codeu/chat/server/Model.java:
```diff
@@ -67,12 +67,23 @@ public int compare(Time a, Time b) {
   private final Store<String, Message> messageByText = new Store<>(STRING_COMPARE);
 
+  private final String version = "1.1";
+  private final Time uptime = Time.now();
```
serverStartTime probably a better variable name.

turtlefarmer commented on src/codeu/chat/util/Time.java:
```diff
@@ -46,6 +47,22 @@ public Time read(InputStream in) throws IOException {
+    long minute = TimeUnit.MILLISECONDS.toMinutes(ms) - (TimeUnit.MILLISECONDS.toHours(ms)* 60);
+    long second = TimeUnit.MILLISECONDS.toSeconds(ms) - (TimeUnit.MILLISECONDS.toMinutes(ms) *60);
+
+    String messageDays = (day != 1) ? day + " days " : day + " day ";
```
does this work?  it would be more compact

day + ((day != 1) ? " days " : "day")

turtlefarmer commented on src/codeu/chat/util/Time.java:
```diff
@@ -46,6 +47,22 @@ public Time read(InputStream in) throws IOException {
+    String messageMinutes = (minute != 1) ? minute + " minutes " : minute + " minute ";
+    String messageSeconds = (second != 0) ? second + " seconds" : second + "second";
+
+    String message = messageDays + messageHours + messageMinutes + messageSeconds;
```
Maybe construct the pieces above without starting/trailing spaces and reconstruct message with a 
String.format.  Starting/trailing spaces usage in string construction is very error prone.

turtlefarmer commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,95 @@
+
+// TOKENIZER
+//
+// This is the class that is used to take in input from the Chat's commandline
```
Use /** for longer comment blocks

turtlefarmer commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,95 @@
+  public Tokenizer(String line) {
+    StringBuilder token = new StringBuilder();
+    boolean inQuotes = false;
+    boolean lookingForEscapable = false; //
```
Empty comment

turtlefarmer commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,95 @@
+// to this one.
+//
+public final class Tokenizer {
+  private LinkedList<String> tokens = new LinkedList<String>();
```
Fun reading:  https://stackoverflow.com/questions/322715/when-to-use-linkedlist-over-arraylist

Not sure if still true with modern day java though.

turtlefarmer commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,95 @@
+        }
+        else {
+          throw new IllegalArgumentException
+          ("Forward slash found before unescapable character.");
```
Usually standard to add at least a couple of spaces when line-wrapping

turtlefarmer commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    addToken(token);
+    token = null;
```
What is the purpose of this?

turtlefarmer commented on src/codeu/chat/util/Tokenizer.java:
```diff
@@ -0,0 +1,95 @@
+        else if (Character.isWhitespace(c) && !inQuotes) {
+          // white space denotes the ending of tokens, meaning the token should
+          // be added to memory
+          addToken(token);
```
Does it make sense for addToken to call clearToken if they're always called in order?

jedouard98 committed:
Fixed Requested Changes (1)

Edited the requested changes from comments from the pull request.
-localized the try catch block
-changed the new network codes to the end
-changed the name of functions in model.java and view.java   to better
show camelCase

jedouard98 committed:
Fixed string formatting

Used a formatter to make code less error prone
